### PR TITLE
Clear private data when closing last private window

### LINKED ISSUE #4125: Private mode data isn't cleared after private windows are closed
Usualy, I open my email in private mode `open -p gmail.com`, so no password will be stored. 

What I expect from a private mode is that all session information (history, commands, cache, passwords, ....) will not be saved OR will be cleared after closing "private windows".  However, I noticed that qutebrowser did not clear passwords and cache of private mode if it is in not totally closed (close all tabs and windows private and non-pivate). If there are non-private windows openend, after closing only private windows, I can get to my mail `open -p gmail.com` without reentring username and password ;( !!!!!!!!!

```
qutebrowser v1.4.1
Git commit: 
Backend: QtWebEngine (Chromium 65.0.3325.230)

CPython: 3.7.0
Qt: 5.11.1
PyQt: 5.11.2

sip: 4.19.12
colorama: 0.3.9
pypeg2: 2.15
jinja2: 2.10
pygments: 2.2.0
yaml: 3.13
cssutils: no
attr: 18.1.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebKitWidgets: yes
pdf.js: no
sqlite: 3.24.0
QtNetwork SSL: OpenSSL 1.1.0h  27 Mar 2018

Style: Qt5CTProxyStyle
Platform: Linux-4.14.56-1-lts-x86_64-with-glibc2.2.5, 64bit
Linux distribution: Artix Linux (unknown)
Frozen: False
Imported from /usr/lib/python3.7/site-packages/qutebrowser
Using Python from /usr/bin/python3
Qt library executable path: /usr/lib/qt/libexec, data path: /usr/share/qt
OS Version: 

--- /etc/lsb-release ---
LSB_VERSION=1.4
DISTRIB_ID=Artix
DISTRIB_RELEASE=rolling
DISTRIB_DESCRIPTION="Artix"

--- /etc/os-release ---
NAME="Artix"
PRETTY_NAME="Artix Linux"
ID=artix
ID_LIKE=artixlinux

Paths:
cache: /home/hk/.cache/qutebrowser
config: /home/hk/.config/qutebrowser
data: /home/hk/.local/share/qutebrowser
runtime: /run/user/1000/qutebrowser
system data: /usr/share/qutebrowser

Uptime: 19:28:23
```

### PR DESCRIPTION
**Pull request by lufte**
This is a work in progress. I still have to:

1. Investigate what happens with webkit backend.
2. Investigate how to clear localStorage data (this doesn't seem to be doing it).
3. Figure out exactly where to put this code. I don't see any other `if objects.backend == usertypes.Backend.QtWebEngine` check in the main window code, so this probably isn't the best place. Let me know what you think.

This fixes #4125.

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/4147)

### CODE REVIEW

**Comment by The-Compiler** [comment]
I'd prefer just throwing the profile away and creating a new profile instead. That'll probably get rid of all data :wink:

**Comment by The-Compiler** [comment]
Doesn't necessarily need to be a slot, but yeah, a single function to call to init a given profile would probably help.

**Comment by lufte** [comment]
Fair enough, plus I see no way of clearing the local storage.

File: qutebrowser/browser/webengine/webenginesettings.py [hunk 1]
```diff
@@ -50,37 +59,41 @@ class _SettingsWrapper:
     For read operations, the default profile value is always used.
     """
 
-    def __init__(self):
-        self._settings = [default_profile.settings(),
-                          private_profile.settings()]
```
Comment (lufte) [inline qutebrowser/browser/webengine/webenginesettings.py]: We can't keep these references to the profiles' settings anymore in case the profiles change.

**Comment by lufte** [comment]
Tests will probably fail with these last changes, I just want to check if this seems like the right approach to you guys. 

Anyway, I'm aware Florian is busy until September so no rush.

File: qutebrowser/browser/webengine/webenginesettings.py [hunk 2]
```diff
@@ -275,23 +288,66 @@ def _update_settings(option):
...
+    try:
+        profile.cookieStore().setCookieFilter(cookies.accept_cookie)
+    except AttributeError:
+        pass
+
+    # Clear visited links on web history clear
+    hist = objreg.get('web-history')
+    hist.history_cleared.connect(profile.clearAllVisitedLinks)
+    hist.url_cleared.connect(lambda url, profile=profile:
+                             profile.clearVisitedLinks([url]))
```
Comment (lufte) [inline qutebrowser/browser/webengine/webenginesettings.py]: Should I call `global_settings.init_settings()` so settings that could've been applied to older profiles are copied to the new one?

Comment (The-Compiler) [reply qutebrowser/browser/webengine/webenginesettings.py]: Hmm, I'm not sure whether it's a good idea to re-apply all settings on existing profiles whenever a new profile is added - but those should definitely be consistent across profiles.

Maybe you could create a temporary `WebEngineSettings` object which only applies the settings to the newly created profile, and after that, `global_settings will take care on updating it for any future changes?

File: qutebrowser/browser/webengine/webenginesettings.py [hunk 3]
```diff
@@ -275,23 +288,66 @@ def _update_settings(option):
...
-
-    default_profile = QWebEngineProfile.defaultProfile()
-    default_profile.setter = ProfileSetter(default_profile)
-    default_profile.setCachePath(
-        os.path.join(standarddir.cache(), 'webengine'))
-    default_profile.setPersistentStoragePath(
-        os.path.join(standarddir.data(), 'webengine'))
-    default_profile.setter.init_profile()
-    default_profile.setter.set_persistent_cookie_policy()
+def create_profile(storage_name=None, private=False):
```
Comment (The-Compiler) [inline qutebrowser/browser/webengine/webenginesettings.py]: Hmm, we start to have more and more stuff in here which isn't really related to settings - maybe we should try moving that out to a `webengineprofile.py` - what do you think?

File: qutebrowser/browser/webengine/webenginesettings.py [hunk 4]
```diff
@@ -275,23 +288,66 @@ def _update_settings(option):
...
+    # Clear visited links on web history clear
+    hist = objreg.get('web-history')
+    hist.history_cleared.connect(profile.clearAllVisitedLinks)
+    hist.url_cleared.connect(lambda url, profile=profile:
+                             profile.clearVisitedLinks([url]))
+    return profile
+
+
+def reset_private_profile():
+    """Replace the private profile with a new instance"""
```
Comment (The-Compiler) [inline qutebrowser/browser/webengine/webenginesettings.py]: nitpick: period :wink:

File: qutebrowser/config/websettings.py [hunk 5]
```diff
@@ -193,6 +193,15 @@ def init(args):
                                 pattern=urlmatch.UrlPattern(pattern))
 
 
+def clear_private_data():
+    """Clear cookies, cache and related data for private browsing sessions"""
```
Comment (The-Compiler) [inline qutebrowser/config/websettings.py]: nitpick: Period :wink:

File: qutebrowser/browser/webengine/webenginesettings.py [hunk 6]
```diff
@@ -302,11 +358,31 @@ def init(args):
...
+    objreg.register('webengine-download-manager', _download_manager)
+
+    global default_profile, private_profile
+    default_profile = create_profile()
+    private_profile = create_profile(private=True)
     config.instance.changed.connect(_update_settings)
 
     global global_settings
-    global_settings = WebEngineSettings(_SettingsWrapper())
+    global_settings = WebEngineSettings(_SettingsWrapper)
```
Comment (The-Compiler) [inline qutebrowser/browser/webengine/webenginesettings.py]: with the `@staticmethod` changes above, this would be an instance rather than a class again.

File: qutebrowser/mainwindow/mainwindow.py [hunk 7]
```diff
@@ -186,7 +187,7 @@ def __init__(self, *, private, geometry=None, parent=None):
             private = True
         else:
             private = bool(private)
-        self._private = private
+        self.private = private
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/mainwindow.py]: nitpick: Please also remove the leading underline in the class docstring (where the `private` attribute is documented).

File: qutebrowser/mainwindow/mainwindow.py [hunk 8]
```diff
@@ -561,6 +562,15 @@ def _do_close(self):
             pass
         objreg.get('session-manager').save_last_window_session()
         self._save_geometry()
+        if (
+                len(objreg.window_registry) > 1
+                and all(window is self and self.private or not window.private
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/mainwindow.py]: I found this a bit hard to follow. Couldn't you just check `self.private and len(window for window in objreg.window_registry.values() if window.private) == 1` or so?

Comment (lufte) [reply qutebrowser/mainwindow/mainwindow.py]: My idea here was to include an extra case: if I'm closing the last private window **but** it is also the last window of any kind, then don't run the cleanup routine. It's just a small improvement to not run unnecesary stuff when shutting down.

Doing `len` instead of `all` would end up in something like:
```
self.private 
and len(objreg.window_registry) > 1
and len(window for window in objreg.window_registry.values() if window.private) == 1
```

which isn't really much shorter, but if you find this clearer I have no problem changing it.

File: qutebrowser/mainwindow/mainwindow.py (hunk 7 above)
Comment (lufte) [reply qutebrowser/mainwindow/mainwindow.py]: :ok_hand:

File: qutebrowser/mainwindow/mainwindow.py (hunk 8 above)
Comment (The-Compiler) [reply qutebrowser/mainwindow/mainwindow.py]: Yeah, I was mainly confused by the `all(window is self and self.private or not window.private ...)` part. Maybe adding an extra set of parentheses would make it a bit clearer, but I would indeed prefer the `len(...)` variant (unless you disagree and find the `all(...)` one much clearer).

Maybe also add a comment like `# Wipe private data if we close the last private window, but there are still other windows` or so - I'm usually not a fan of comments duplicating the code's logic, but here I think it'd be worth it.

**Comment by lufte** [comment]
Hey, thanks for reviewing. I feel most of the comments are related to the last commit 99a296154f48a8c3ffd91f3e8b99e73b76b1d272, of which I'm also a bit dubious and also it's not completely related to the issue I was solving. Do you think it would make sense to merge the PR without that commit? It definitely fixes the original issue and we can open another issue to discuss how to handle the multiple profiles, being that #4125 is priority 0 and all.

I'll take a look and review your other comments in the meantime anyway.

File: qutebrowser/config/websettings.py (hunk 5 above)
Comment (lufte) [reply qutebrowser/config/websettings.py]: :ok_hand:

File: qutebrowser/browser/webengine/webenginesettings.py (hunk 1 above)
Comment (lufte) [reply qutebrowser/browser/webengine/webenginesettings.py]: That sounds good.

File: qutebrowser/browser/webengine/webenginesettings.py (hunk 2 above)
Comment (lufte) [reply qutebrowser/browser/webengine/webenginesettings.py]: I'm not really familiar with all the "settings" part of the code but I think I get what you mean. I'll take care of it in the next commits.

File: qutebrowser/browser/webengine/cookies.py [hunk 9]
```diff
@@ -34,15 +34,3 @@ def _accept_cookie(request):
         return False
     else:
         raise utils.Unreachable
-
-
-def install_filter(profile):
```
[the remark this answers is not in the document]
Comment (lufte) [reply qutebrowser/browser/webengine/cookies.py]: Hmm, looking back now it seems I went a little crazy with changing code based exclusively on personal preference :sweat_smile: 

I think my reasoning was "this made me think it was going to install **the profile on something** but was actually the other way around, so why not put this code directly in the profile init code?". Anyway, doesn't make much sense when looking at it currently, I'll change it back.

File: qutebrowser/browser/webengine/webenginesettings.py [hunk 10]
```diff
@@ -50,37 +59,41 @@ class _SettingsWrapper:
     For read operations, the default profile value is always used.
     """
 
-    def __init__(self):
-        self._settings = [default_profile.settings(),
-                          private_profile.settings()]
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (lufte) [reply qutebrowser/browser/webengine/webenginesettings.py]: Alright, but are you sure you don't like them? I think they keep the code cleaner and also they are definitely faster, even when there is no `__init__` method:
```python
Python 3.5.3 (default, Sep 27 2018, 17:25:39)
[GCC 6.3.0 20170516] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> class MyClass:
...  def normal(self):
...   return
...  @staticmethod
...  def static():
...   return
...
>>> import timeit
>>> timeit.timeit('MyClass().normal()', globals=globals())
0.24852409300365252
>>> timeit.timeit('MyClass.static()', globals=globals())
0.15611451499717077
```

**Comment by The-Compiler** [comment]
I'm fine with both I guess - I agree it'd probably have made sense to split it up into two PRs, but since I already reviewed it, I'd say it's probably easier to keep it now.

**Comment by lufte** [comment]
Hey, sorry I kind of abandoned this. I'm gonna revert commit 99a296154f48a8c3ffd91f3e8b99e73b76b1d272 so this is strictly a fix for #4125, plus I can't promise I'll work on the other stuff discussed here. So, to refresh the memory, this pr...
1- Will delete and create a new private profile when the last private window is closed in QtWebEngine.
2- Will clear cookies when the last private window is closed in QtWebkit.
3- Will *not* implement a function for creating new profiles.

**Comment by jgkamat** [comment]
Could you merge master into this branch? There's a couple small merge conflicts.

It also looks like `test_turning_off_javascript_with_private_browsing` is failing, but I haven't triaged that yet. The code looks very solid though :)

**Comment by lufte** [comment]
Thanks. Ok great, I'll try to tidy it up during the weekend.

**Comment by lufte** [comment]
I took care of the build errors that I think were caused by my changes. I also added two new tests and fixed a bug that was indeed found by one of the test cases.

Let me know if it looks good to you @jgkamat

**Comment by lufte** [comment]
@jgkamat Hey, any clue on why my tests only fail in the Windows build? I'm not really familiar with the way these fixture tests are written but I tried to mimic what other tests were doing and they work on Travis.

**Comment by toofar** [comment]
In the logs there is "schedule removal" logs but no "removed logs" https://github.com/qutebrowser/qutebrowser/blob/master/qutebrowser/utils/objreg.py#L124
Not sure if things are just taking a little longer or what. Maybe telling it to wait for "removed: last-visible-main-window" would help. I wonder if it would reproduce in a VM.

**Comment by lufte** [comment]
Thanks, I see you already solved it in an pull request of yours. Is this the command I should add after you merge that? 
https://github.com/toofar/qutebrowser/blob/feat/window_undo/tests/end2end/features/tabs.feature#L798

**Comment by jgkamat** [comment]
Javier Ayres writes:

> @jgkamat Hey, any clue on why my tests only fail in the Windows build? I'm 
> not really familiar with the way these fixture tests are written but I tried 
> to mimic what other tests were doing and they work on Travis.

This is definitely an issue that needs to be fixed. As toofar said it's 
probably a race condition with closing the window. I wonder if we can 
reproduce this on master too, and this test just exposed that :).

**Comment by toofar** [comment]
> Thanks, I see you already solved it in an pull request of yours. Is this the command I should add after you merge that?
> https://github.com/toofar/qutebrowser/blob/feat/window_undo/tests/end2end/features/tabs.feature#L798

Yeah that and the `wait for "removed: tabbed-browser"` one.

> I wonder if we can reproduce this on master too

It was happening reliably on appveyor but I couldn't reproduce it on a windows VM, even with one core. If it is reproducable in other situations (like on a potato running a userscript that closes windows etc) the proper fix might be to remove some objects from objreg immediately. Like the keys that `get_window()` for IPC messages uses.

File: qutebrowser/mainwindow/mainwindow.py [hunk 11]
```diff
@@ -582,6 +582,19 @@ def _do_close(self):
             pass
         objreg.get('session-manager').save_last_window_session()
         self._save_geometry()
+
+        # Wipe private data if we close the last private window, but there are
+        # still other windows
+        if (
+                self.private and
+                len(objreg.window_registry) > 1 and
```
[the remark this answers is not in the document]
Comment (lufte) [reply qutebrowser/mainwindow/mainwindow.py]: Not really, line 590 asks for more than one window while line 591 asks for exactly one private window. I want to clear the private profile if we are closing the last private window but there's still another window around. If we are closing the last window overall, then the private profile is going to be cleared anyway; no need to do it here.

File: qutebrowser/config/websettings.py [hunk 12]
```diff
@@ -198,6 +198,15 @@ def init(args: argparse.Namespace) -> None:
                                 pattern=urlmatch.UrlPattern(pattern))
 
 
+def clear_private_data() -> None:
+    """Clear cookies, cache and related data for private browsing sessions."""
+    if objects.backend == usertypes.Backend.QtWebEngine:
+        from qutebrowser.browser.webengine import webenginesettings
+        webenginesettings.init_private_profile()
+    else:
```
[the remark this answers is not in the document]
Comment (lufte) [reply qutebrowser/config/websettings.py]: Agreed, although I just copied the pattern used in another two functions in that file: check for QtWebEngine, if not assume QtWebKit. Since I'm here I made the same adjustment to those functions as well.

Thanks for your comments!

**Comment by The-Compiler** [comment]
Almost on its 2nd birthday I finally merged this! :tada:

As usual, with a couple of additional changes and cleanups:

- e52cbedef Merge remote-tracking branch 'origin/pr/4147' into last-private
- 735ab6b42 Fix self._private access
- bb9f4d179 Fix lint
- 95186a033 Fix clearing cookie jar on QtWebKit
- ba86a040d Simplify _SettingsWrapper
- e2a07148e Unify backend conditionals
- 6f554b4e3 Rename self.private to self.is_private
- 35d116dd4 Update changelog

Thanks @lufte for the PR and @jgkamat @toofar for the reviews/comments!

**Comment by lufte** [comment]
Haha, awesome! She was so close to start attending kindergarten though =D
